### ISSUE TITLE
QC Failures: Nifty Header & Sequence Support

### ISSUE THREAD

**Issue by m-reuter** [Deep-MI/FastSurfer#431]
I made the changes to --user 1001:1001 and it kinda worked but now I have error "FastSurfer asegdkt segmentation failed".

C:\Windows\System32>docker run -v C:/Users/Ernes/my_mri_data:/data -v C:/Users/Ernes/my_fastsurfer_analysis:/output -v C:/Users/Ernes/my_fs_license_dir:/fs_license --rm --user 1001:1001 deepmi/fastsurfer:cpu-latest --fs_license /fs_license/license.txt --t1 /data/subjectX/H1.nii.gz --device cpu --sid subjectX --sd /output --parallel
Version: 2.2.0+9f37d02
Sun Jan 14 08:23:58 UTC 2024

python-s /fastsurfer/FastSurferCNN/run_prediction.py --t1 /data/subjectX/H1.nii.gz --asegdkt_segfile /output/subjectX/mri/aparc.DKTatlas+aseg.deep.mgz --conformed_name /output/subjectX/mri/orig.mgz --brainmask_name /output/subjectX/mri/mask.mgz --aseg_name /output/subjectX/mri/aseg.auto_noCCseg.mgz --sid subjectX --seg_log /output/subjectX/scripts/deep-seg.log --vox_size min --batch_size 1 --viewagg_device auto --device cpu
[INFO: run_prediction.py:  546]: Checking or downloading default checkpoints ...
[INFO: common.py:  111]: Using device: cpu
[INFO: run_prediction.py:  234]: Running view aggregation on cpu
[INFO: inference.py:  200]: Loading checkpoint /fastsurfer/checkpoints/aparc_vinn_coronal_v2.0.0.pkl
[INFO: inference.py:  200]: Loading checkpoint /fastsurfer/checkpoints/aparc_vinn_sagittal_v2.0.0.pkl
[INFO: inference.py:  200]: Loading checkpoint /fastsurfer/checkpoints/aparc_vinn_axial_v2.0.0.pkl
[INFO: common.py:  820]: Single subject with absolute file path for input.
[INFO: common.py:  833]: No subjects directory specified, but the parent directory of the output file /output/subjectX/mri/aparc.DKTatlas+aseg.deep.mgz is 'mri', so we are assuming this is the 'mri' folder in the subject directory.
[INFO: common.py:  875]: Analyzing single subject /data/subjectX/H1.nii.gz
[INFO: common.py:  970]: Output will be stored in Subjects Directory: /output
[INFO: run_prediction.py:  310]: Successfully loaded image from /data/subjectX/H1.nii.gz.
[INFO: run_prediction.py:  418]: Output image directory 001.mgz does not exist. Creating it now...
Traceback (most recent call last):
  File "/fastsurfer/FastSurferCNN/run_prediction.py", line 561, in <module>
    for subject, (orig_img, data_array) in iter_subjects:
  File "/fastsurfer/FastSurferCNN/run_prediction.py", line 462, in pipeline_conform_and_save_orig
    yield subject, self.conform_and_save_orig(subject)
  File "/fastsurfer/FastSurferCNN/run_prediction.py", line 316, in conform_and_save_orig
    if not conf.is_conform(
  File "/fastsurfer/FastSurferCNN/data_loader/conform.py", line 625, in is_conform
    iaffine = img.affine[0:3, 0:3] - LIA_affine * (
ValueError: operands could not be broadcast together with shapes (3,3) (4,)
ERROR: FastSurfer asegdkt segmentation failed.

_Originally posted by @KontorErnestKissi in https://github.com/Deep-MI/FastSurfer/issues/428#issuecomment-1890884814_

**Comment by m-reuter** [comment]
what resolution (voxel sizes and dimensions) is your subjectX/H1.nii.gz image? With FreeSurfer you could run `mri_info` on this and post the result. 

In your case the izoom seems to be 4 dimensional, a normal image should only have three dimensions. 
https://github.com/Deep-MI/FastSurfer/blob/bed50c3431c5b8b0cff73469527dfb2ab8e274ac/FastSurferCNN/data_loader/conform.py#L626

**Comment by KontorErnestKissi** [comment]
INFO: using NIfTI-1 qform
Volume information for H1.nii.gz
          type: nii
    dimensions: 192 x 192 x 160
   voxel sizes: 1.250000, 1.250000, 1.249826
          type: SHORT (4)
           fov: 240.000
           dof: 1
        xstart: -120.0, xend: 120.0
        ystart: -120.0, yend: 120.0
        zstart: -100.0, zend: 100.0
            TR: 0.00 msec, TE: 0.00 msec, TI: 0.00 msec, flip angle: 0.00 degrees
       nframes: 1
       PhEncDir: UNKNOWN
       FieldStrength: 0.000000
ras xform present
    xform info: x_r =   1.0000, y_r =   0.0000, z_r =   0.0000, c_r =     0.0000
              : x_a =   0.0000, y_a =   1.0000, z_a =   0.0000, c_a =     1.2500
              : x_s =   0.0000, y_s =   0.0000, z_s =   1.0000, c_s =     1.2498
Orientation   : RAS
Primary Slice Direction: axial

voxel to ras transform:
                1.2500   0.0000   0.0000  -120.0000
                0.0000   1.2500   0.0000  -118.7500
                0.0000   0.0000   1.2498   -98.7363
                0.0000   0.0000   0.0000     1.0000

**Comment by m-reuter** [comment]
Hmm, that looks OK to me. Strange. We can take a closer look at the code on Monday. We may need to ask for an example image (can be completely zero, just the header should be the same as this one and it should also fail for you at the same place). That could help for debugging this. 

In the meantime, you can try running free surfers `mri_convert H1.nii.gz H1.mgz`and then passing the mgz into fastsurfer. Or directly using FreeSurfer to conform the image `mri_convert -c H1.nii.gz H1-conform.mgz`

**Comment by KontorErnestKissi** [comment]
Thank you for your prompt response. I appreciate it. 
I tried this one : mri_convert -c H1.nii.gz H1-conform.mgz and it worked . But it has been 6 hours now and  it is not done running .

 I thought it was supposed to take 60-90 mins for surface reconstruction.This is where it has run to and it has been there since:

 RUNNING HEMIs in PARALLEL !!!


RunBatchJobs: Logfile: /output/subjectX/scripts/recon-surf.log
RunBatchJobs: CMDF: /output/subjectX/scripts/lh.processing.cmdf
RunBatchJobs: CMDF: /output/subjectX/scripts/rh.processing.cmdf
Waiting for PID 1282 of (1282 1285) to complete...

**Comment by dkuegler** [comment]
It should finish way faster than 6 hours.
Can you attach all files in `C:/Users/Ernes/my_fastsurfer_analysis/subjectX/scripts/` ?
This should have the relevant log files for us to see what is going on. Alternatively, remove the `--parallel` from your command and that then puts all outputs in the command line. But again, that is different and "new issue".


Maybe this has something to do with the fact that this image is >1.0mm in all directions?

what is definitely a bit weird is `np.max(np.abs(izoom - conformed_vox_size) < eps)` the brackets seem off... it should be `np.max(np.abs(izoom - conformed_vox_size)) < eps`. 

But that does not change the fact, that very likely `nibabel` returns 4 values in `img.header.get_zooms()`. And that specific error we will only be able to debug if we have the exact same header as it is unclear, if that is a bug in `nibabel` or FastSurfer.

**Comment by KontorErnestKissi** [comment]
These are the files:
[build.log](https://github.com/Deep-MI/FastSurfer/files/13932834/build.log)
[build-stamp.txt](https://github.com/Deep-MI/FastSurfer/files/13932835/build-stamp.txt)
[deep-seg.log](https://github.com/Deep-MI/FastSurfer/files/13932836/deep-seg.log)
[defect2seg.log](https://github.com/Deep-MI/FastSurfer/files/13932837/defect2seg.log)
[lastcall.build-stamp.txt](https://github.com/Deep-MI/FastSurfer/files/13932838/lastcall.build-stamp.txt)
[patchdir.txt](https://github.com/Deep-MI/FastSurfer/files/13932839/patchdir.txt)
[ponscc.cut.log](https://github.com/Deep-MI/FastSurfer/files/13932840/ponscc.cut.log)
[recon-all.log](https://github.com/Deep-MI/FastSurfer/files/13932841/recon-all.log)
[recon-all-status.log](https://github.com/Deep-MI/FastSurfer/files/13932842/recon-all-status.log)
[recon-surf.log](https://github.com/Deep-MI/FastSurfer/files/13932843/recon-surf.log)
[unknown-args.txt](https://github.com/Deep-MI/FastSurfer/files/13932844/unknown-args.txt)

**Comment by dkuegler** [comment]
There is something happening in the right hemisphere: The first fail seems to be in `spherical_project.py`.
From `recon-surf.log`:
```
Command python-s /fastsurfer/recon_surf/spherically_project.py -i /output/subjectX/surf/rh.smoothwm.nofix -o /output/subjectX/surf/rh.qsphere.nofix failed.
```

I'll talk about this with @m-reuter tomorrow.

**Comment by dkuegler** [comment]
As for the original error that is at the top of the issue: I think we need your header (from your image) to reproduce that error...
If you are able to share that original `H1.nii.gz` file, please get in touch via email. I don't partricularly know how to make sure you will send the exact header otherwise. Many other conversion tools might modify the header (as `mri_info` does not seem to have the issue with the four dimensions).

**Comment by KontorErnestKissi** [comment]
You can find the image file below:
***deleted***

Thank you .

**Comment by dkuegler** [comment]
I have received the file, I will take a look tomorrow. I also removed the link from your post so this file does not leak into the general public.

**Comment by KontorErnestKissi** [comment]
Thank you very much.

**Comment by dkuegler** [comment]
I had a look at the file you provided and it is not oriented as the header suggests.
Overall, the headers seem pretty inconsistent to even corrupted.

I think the header says the image is stored as RAS, while in fact it might be stored as LIA.

**Comment by m-reuter** [comment]
When opening the image in FreeView it is not oriented correctly but flipped. This indicates a corrupt header. Furthermore, the image is 4D but with a single frame (so the 4th dimension has size 1). That is also unusual. Probably something went wrong when converting this image from DICOM to nifti? 

The failure later in the pipeline and the long runtime are probably caused by the incorrect orientation.

**Comment by KontorErnestKissi** [comment]
So how do you suggest I correct this? Because I used the mri_convert -c H1.nii.gz H1-conform.mgz but still it took a longer time. I have more data so I will try on different images and see the outcome. Thanks

**Comment by m-reuter** [comment]
The mri_convert can deal with the 4-dimensional single-frame input but it does not (and cannot) correct the header orientation. There is actually no way of telling left from right just from the image. If this information get's corrupted on the way from the scanner to you, you need to investigate each step and ensure that no information gets lost in any of the conversion steps. Talk to who provided the images and let them know that orientation information got lost somewhere.

**Comment by KontorErnestKissi** [comment]
I will do that.  Okay, thank you so much. You and your team are awesome.

**Comment by KontorErnestKissi** [comment]
![image](https://github.com/Deep-MI/FastSurfer/assets/73202044/1e002bfa-b10b-445f-a584-e97a28cfd944)

[build.log](https://github.com/Deep-MI/FastSurfer/files/13983479/build.log)
[build-stamp.txt](https://github.com/Deep-MI/FastSurfer/files/13983481/build-stamp.txt)
[deep-seg.log](https://github.com/Deep-MI/FastSurfer/files/13983482/deep-seg.log)
[lastcall.build-stamp.txt](https://github.com/Deep-MI/FastSurfer/files/13983483/lastcall.build-stamp.txt)
[patchdir.txt](https://github.com/Deep-MI/FastSurfer/files/13983484/patchdir.txt)
[ponscc.cut.log](https://github.com/Deep-MI/FastSurfer/files/13983485/ponscc.cut.log)
[recon-all.log](https://github.com/Deep-MI/FastSurfer/files/13983486/recon-all.log)
[recon-all-status.log](https://github.com/Deep-MI/FastSurfer/files/13983487/recon-all-status.log)
[recon-surf.log](https://github.com/Deep-MI/FastSurfer/files/13983488/recon-surf.log)
[unknown-args.txt](https://github.com/Deep-MI/FastSurfer/files/13983489/unknown-args.txt)
![image](https://github.com/Deep-MI/FastSurfer/assets/73202044/a2222f22-e57c-42f8-8e5c-0903f615e677)



Now I am using an image with the orientation. The recon-all seams to be done but the segmentation has some error. Above is the screenshot of the error, stat file , screenshot of image and the script files.The stat file also looks  weird though the recon all is done.

**Comment by dkuegler** [comment]
Looking at the statsfile (is this aseg.stats or aseg+dkt.stats?) looks like the segmentation failed.

In `deep-seg.log` you can read, that the automatic QC found the segmentation did not work.
> [WARNING: run_prediction.py:  621]: Total segmentation volume is too small. Segmentation may be corrupted.

So it looks like as if already in the very first step you get errors. There are a couple of reasons this might happen. Some sort of problem with the data is the most likely. 

It is a bit hard to say from the image screenshots (the quality is not great), but this does not look like a T1w image to me. FastSurfer needs a T1w image.

**Comment by m-reuter** [comment]
Yes, FastSurfer works on T1 weighted MRI, such as a MPRAGE or multi-echo MPRAGE (recommended) in Siemens and product MPRAGE in Philips, with isotropic voxel sizes between 0.7 and 1mm, and probably also a little outside this range.  Also GE product sequence IR-FSPGR should work. Generally all ADNI sequences should be fine (see https://adni.loni.usc.edu/wp-content/themes/freshnews-dev-v2/documents/mri/ADNI_MRI_Methods_Non-ADNI_Studies.pdf ) .

**Comment by KontorErnestKissi** [comment]
It is T1w FLASH image in Siemens

**Comment by dkuegler** [comment]
We are not certain Fastsurfer supports flash sequences. We are currently going through our datasets to see whether we have flash images in the test and/or the training datasets

**Comment by KontorErnestKissi** [comment]
It worked fine. It took exactly 90 mins. I was able to process 2 images but the results were exactly the same. But in freesurfer, the results were different. So I am now using the freeSurfer for my project. 
Perhaps, because I am working with Macromolecular Proton Fraction(MPF) images which is a new modality. It would have been great if they were working in fastSurfer.
Thanks for the supports. I really appreciate it. All the best.

**Comment by dkuegler** [comment]
To clarify on the previously unanswered question of support for FLASH sequences.
A quick qualitative assessment showed, that FastSurfer 2.2 does properly segment T1-FLASH images, but results seem a bit less robust (which makes sense, as there are no FLASH T1w images in the training dataset).
We did not quantitatively evaluate the segmentation, but results look generally okay. Users should still be careful, but given QC, results should be valid.
Segmentation results for the cerebellum (CerebNet) seem also fine.
